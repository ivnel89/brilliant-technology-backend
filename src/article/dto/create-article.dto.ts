import { ApiProperty } from "@nestjs/swagger";
import { IsString, IsUUID } from "class-validator";

export class CreateArticleDto {
    @ApiProperty()
    @IsString()
    content: string;
  
    @ApiProperty()
    @IsString()
    @IsUUID()
    authorId: string;
}
