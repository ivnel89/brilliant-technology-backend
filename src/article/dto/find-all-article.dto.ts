import { ApiProperty } from "@nestjs/swagger";
import { IsNumber, IsOptional, IsUUID, Max } from "class-validator";

export class FindAllArticleDto {
    @IsOptional()
    @IsNumber()
    @ApiProperty({
      required: false
    })
    offset?: number;

    @IsOptional()
    @IsNumber()
    @Max(250)
    @ApiProperty({
      required: false
    })
    limit?: number;

    @IsOptional()
    @ApiProperty({
      required: false
    })
    @IsUUID()
    requesterId?: string;
  }
  