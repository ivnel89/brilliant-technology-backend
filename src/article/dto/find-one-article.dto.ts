import { ApiProperty } from "@nestjs/swagger";
import { IsOptional, IsUUID } from "class-validator";

export class FindOneArticleDto {
    @IsOptional()
    @ApiProperty({
      required: false
    })
    @IsUUID()
    requesterId?: string;
  }
  