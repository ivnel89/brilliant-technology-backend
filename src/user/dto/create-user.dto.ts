import { ApiProperty } from "@nestjs/swagger";
import { IsOptional, IsString, IsUrl, Length } from "class-validator";

export class CreateUserDto {
  @ApiProperty()
  @IsString()
  @Length(1,256)
  firstName: string;

  @ApiProperty()
  @IsString()
  @Length(1,256)
  lastName: string;

  @ApiProperty()
  @IsUrl()
  @Length(1,2048)
  displayPicture: string;
}
