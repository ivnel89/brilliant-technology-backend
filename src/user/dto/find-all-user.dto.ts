import { ApiProperty } from "@nestjs/swagger";
import { IsNumber, IsOptional, Max } from "class-validator";

export class FindAllUserDto {
    @IsOptional()
    @IsNumber()
    @ApiProperty()
    offset?: number;

    @IsOptional()
    @IsNumber()
    @Max(250)
    @ApiProperty()
    limit?: number;
  }
  