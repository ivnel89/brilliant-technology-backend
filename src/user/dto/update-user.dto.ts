import { ApiProperty } from "@nestjs/swagger";
import { IsOptional, IsString, IsUrl, Length } from "class-validator";

export class UpdateUserDto {
    @ApiProperty()
    @IsString()
    @IsOptional()
    @Length(1,256)
    firstName?: string;
  
    @ApiProperty({
        required: false
      })
    @IsString()
    @IsOptional()
    @Length(1,256)
    lastName?: string;
  
    @ApiProperty({
        required: false
      })
    @IsUrl()
    @IsOptional()
    @Length(1,2048)
    displayPicture?: string;

}
