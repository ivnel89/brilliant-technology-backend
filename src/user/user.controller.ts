import { Controller, Get, Post, Body, Patch, Param, Delete, Query } from '@nestjs/common';
import { UserService } from './user.service';
import { CreateUserDto } from './dto/create-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';
import { FindAllUserDto } from './dto/find-all-user.dto';
import { UserContractMapper } from './mapper/user.mapper';
import { UserContract } from './contract/user.contract';

@Controller('user')
export class UserController {
  constructor(private readonly userService: UserService) {}
  private userContractMapper = new UserContractMapper();

  @Post()
  async create(@Body() createUserDto: CreateUserDto): Promise<UserContract> {
    const user = await this.userService
      .create(createUserDto);
    return this.userContractMapper.build(user);
  }

  @Get()
  async findAll(@Query() findAllUserDto?: FindAllUserDto): Promise<Array<UserContract>> {
    const users = await this.userService
      .findAll(findAllUserDto);
    return this.userContractMapper.buildArray(users);
  }

  @Get(':id')
  async findOne(@Param('id') id: string): Promise<UserContract> {
    const user = await this.userService
      .findOne(id);
    return this.userContractMapper.build(user);
  }

  @Patch(':id')
  async update(@Param('id') id: string, @Body() updateUserDto: UpdateUserDto): Promise<UserContract> {
    const user = await this.userService
      .update(id, updateUserDto);
    return this.userContractMapper.build(user);
  }

  @Delete(':id')
  remove(@Param('id') id: string) {
    return this.userService.remove(id);
  }
}
