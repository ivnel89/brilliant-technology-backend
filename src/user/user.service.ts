import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import {  Repository, UpdateResult } from 'typeorm';
import { CreateUserDto } from './dto/create-user.dto';
import { FindAllUserDto } from './dto/find-all-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';
import { User } from './entities/user.entity';

@Injectable()
export class UserService {
  constructor(
    @InjectRepository(User)
    private userRepository: Repository<User>,
  ) {}

  async create(createUserDto: CreateUserDto): Promise<User> {
    return this.userRepository.save(
      new User(
        createUserDto?.firstName,
        createUserDto?.lastName,
        createUserDto?.displayPicture,
      ),
    );
  }

  findAll(findAllUserDto: FindAllUserDto): Promise<Array<User>> {
    const { limit = 10, offset = 0} = findAllUserDto
    return this.userRepository.find({
      order: {
        lastModifiedDate: 'ASC',
      },
      skip: offset,
      take: limit,
    });
  }

  findOne(id: string): Promise<User> {
    return this.userRepository.findOne({
      where: {
        id,
      },
    });;
  }

  async update(id: string, updateUserDto: UpdateUserDto): Promise<User> {
    const user = await this.findOne(id);
    Object.keys(updateUserDto).forEach( key => {
      if(updateUserDto[key]){
        user[key] = updateUserDto[key];
      }
    })
    return this.userRepository.save(user);
  }

  remove(id: string): Promise<UpdateResult> {
    return this.userRepository.softDelete({ id });
  }
}
