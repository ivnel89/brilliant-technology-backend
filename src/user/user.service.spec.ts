import { Test, TestingModule } from '@nestjs/testing';
import { getRepositoryToken } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { CreateUserDto } from './dto/create-user.dto';
import { FindAllUserDto } from './dto/find-all-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';
import { User } from './entities/user.entity';
import { UserService } from './user.service';

const mockUser = new User(
  "Foo",
  "Bar",
  "www.example.com/img.png"
)

const UserRepository = {
  provide: getRepositoryToken(User),
  useValue: {
    save: jest.fn().mockImplementation((user: User) => user),
    findOne: jest.fn().mockResolvedValue(mockUser),
    create: jest.fn(),
    softDelete: jest.fn().mockResolvedValue("Deleted"),
    find: jest.fn().mockResolvedValue([mockUser]),
  },
};

describe('UserService', () => {
  let service: UserService;
  let repository: Repository<User>;
  let repositorySaveSpy, repositoryFindSpy, repositoryFindOneSpy, repositorySoftDeleteSpy;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        UserService,
        UserRepository
      ],
    }).compile();
    service = module.get<UserService>(UserService);
    repository = module.get<Repository<User>>(
      getRepositoryToken(User),
    );
    repositorySaveSpy = jest.spyOn(repository,'save');
    repositoryFindSpy = jest.spyOn(repository,'find');
    repositoryFindOneSpy = jest.spyOn(repository,'findOne');
    repositorySoftDeleteSpy= jest.spyOn(repository,'softDelete');
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('create', () => {
    
    const createUserDto = new CreateUserDto();
    createUserDto.displayPicture = "www.example.com/image.png";
    createUserDto.firstName = "Foo";
    createUserDto.lastName = "Bar";

    const saveMockError = "saveMockError"

    it('should call repository\'s save method with right params', async () => {
      await expect(service.create(createUserDto))
        .resolves.toMatchSnapshot()
        .then(() => {
          expect(repositorySaveSpy).toBeCalledWith(
            new User(
              createUserDto.firstName,
              createUserDto.lastName,
              createUserDto.displayPicture,
            ),
          );
        });
    })

    it('should throw an error when save throws an error', async () => {
      repositorySaveSpy.mockRejectedValueOnce(saveMockError);
      await expect(service.create(createUserDto)).rejects.toBe(saveMockError);
    })

  })

  describe('findAll', () => {
    const findAllUserDto = new FindAllUserDto();
    findAllUserDto.limit = 50;
    findAllUserDto.offset = 10;

    const findMockError = "findMockError";

    it('should call repository\'s find method with right params', async () => {
      await expect(service.findAll(findAllUserDto))
        .resolves.toMatchSnapshot()
        .then(() => {
          expect(repositoryFindSpy).toBeCalledWith(
            {
              order: {
                lastModifiedDate: 'ASC',
              },
              skip: findAllUserDto.offset,
              take: findAllUserDto.limit,
            },
          );
        });
    })

    it('should throw an error when find throws an error', async () => {
      repositoryFindSpy.mockRejectedValueOnce(findMockError);
      await expect(service.findAll(findAllUserDto))
        .rejects.toBe(findMockError)
    })

  })

  describe('findOne', () => {
    
    const mockUserId = "mockUserId";
    const findOneMockError = "findOneMockError";

    it('should call repository\'s findOne method with right params', async () => {
      await expect(service.findOne(mockUserId))
        .resolves.toMatchSnapshot()
        .then(() => {
          expect(repositoryFindOneSpy).toBeCalledWith({
            where: {
              id: mockUserId,
            },
          });
        });
    })

    it('should throw an error when findOne throws an error', async () => {
      repositoryFindOneSpy.mockRejectedValueOnce(findOneMockError);
      await expect(service.findOne(mockUserId))
        .rejects.toBe(findOneMockError)
    })

  })

  describe('update', () =>  {
    const mockUserId = "mockUserId";
    const updateUserDto = new UpdateUserDto();
    updateUserDto.lastName = "Far";
    const saveMockError = "saveMockError"

    beforeEach(() => {
      repositoryFindOneSpy.mockResolvedValueOnce(mockUser);
    })

    it('should call repository\'s save & findOne method with right params', async () => {      
      await expect(service.update(mockUserId,updateUserDto))
        .resolves.toMatchSnapshot()
        .then(() => {
          expect(repositoryFindOneSpy).toBeCalledWith({
            where: {
              id: mockUserId,
            },
          });
          expect(repositorySaveSpy).toBeCalledWith(
            new User(
              mockUser.firstName,
              updateUserDto.lastName,
              mockUser.displayPicture,
            ),
          );
        });
    })

    it('should throw an error when save throws an error', async () => {
      repositorySaveSpy.mockRejectedValueOnce(saveMockError);
      await expect(service.update(mockUserId,updateUserDto)).rejects.toBe(saveMockError);
    })
  })

  describe('remove', () => {
    const mockUserId = "mockUserId";
    const softDeleteMockError = "softDeleteMockError"
    it('should call repository\'s softDelete method with right params', async () => {      
      await expect(service.remove(mockUserId))
        .resolves.toMatchSnapshot()
        .then(() => {
          expect(repositorySoftDeleteSpy).toBeCalledWith({
            id: mockUserId,
          });
        });
    })
    it('should throw an error when softDelete throws an error', async () => {
      repositorySoftDeleteSpy.mockRejectedValueOnce(softDeleteMockError);
      await expect(service.remove(mockUserId)).rejects.toBe(
        softDeleteMockError,
      );
    });
  })

});
