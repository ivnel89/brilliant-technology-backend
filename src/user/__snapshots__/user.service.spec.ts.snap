// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserService create should call repository's save method with right params 1`] = `
User {
  "displayPicture": "www.example.com/image.png",
  "firstName": "Foo",
  "isActive": true,
  "lastName": "Bar",
}
`;

exports[`UserService findAll should call repository's find method with right params 1`] = `
Array [
  User {
    "displayPicture": "www.example.com/img.png",
    "firstName": "Foo",
    "isActive": true,
    "lastName": "Bar",
  },
]
`;

exports[`UserService findOne should call repository's findOne method with right params 1`] = `
User {
  "displayPicture": "www.example.com/img.png",
  "firstName": "Foo",
  "isActive": true,
  "lastName": "Bar",
}
`;

exports[`UserService remove should call repository's softDelete method with right params 1`] = `"Deleted"`;

exports[`UserService update should call repository's save & findOne method with right params 1`] = `
User {
  "displayPicture": "www.example.com/img.png",
  "firstName": "Foo",
  "isActive": true,
  "lastName": "Far",
}
`;
